### PR TITLE
Poll async service deletions

### PR DESCRIPTION
**Pull request by petko-boyadzhiev**
#### Description: 
Currently, all processes that include deletion of services only just trigger the deletion without waiting for those deletions to finish. This could cause problems if the user then attempts to redeploy the same MTA immediately after the process finishes. 

This pull request adopts an already established (in the current BPMN) polling mechanism for waiting for the actual service deletion.

### CODE REVIEW

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/clients/EventsGetter.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import org.cloudfoundry.client.lib.util.CloudEntityResourceMapper;
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.RestTemplate;
+
+@Component
+public class EventsGetter extends CustomControllerClient {
+
+	private static final String GUID = "guid";
+	private static String EVENTS_URL = "/v2/events";
+	private CloudEntityResourceMapper resourceMapper = new CloudEntityResourceMapper();
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/clients/EventsGetter.java]: The CloudEntityResourceMapper is being used a lot in the other implementations of the CustomControllerClient. Could we make the logic generic?

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/clients/ServiceInstanceGetter.java [hunk 2]
```diff
@@ -53,13 +59,11 @@ public ServiceInstanceGetter(RestTemplateFactory restTemplateFactory) {
...
     }
 
-    @SuppressWarnings("unchecked")
     private Map<String, Object> getCloudServiceInstance(Map<String, Object> serviceInstancesResponse) {
         validateServiceInstanceResponse(serviceInstancesResponse);
         List<Map<String, Object>> cloudServiceInstanceResources = getResourcesFromResponse(serviceInstancesResponse);
         if (!cloudServiceInstanceResources.isEmpty()) {
-            return (Map<String, Object>) cloudServiceInstanceResources.get(0)
-                .get("entity");
+            return cloudServiceInstanceResources.get(0);
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/clients/ServiceInstanceGetter.java]: By this, you are returning the whole response from the Controller. Why not getting just the entity?
Just to remote the SuppressWarning?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/Constants.java [hunk 3]
```diff
@@ -99,6 +99,7 @@
     public static final String VAR_APPS_TO_RESTART_COUNT = "appsToRestartCount";
     public static final String VAR_APPS_TO_UNDEPLOY = "appsToUndeploy";
     public static final String VAR_SERVICES_TO_DELETE = "servicesToDelete";
+    public static final String VAR_SERVICES_GUID = "servicesGuids";
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/Constants.java]: TYPO VAR_SERVICES_GUID"s"

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/message/Messages.java [hunk 4]
```diff
@@ -135,6 +135,9 @@
     public static final String ERROR_UPDATING_OPTIONAL_SERVICE = "Error updating optional service \"{0}\" from offering \"{1}\" and plan \"{2}\": {3}";
     public static final String ERROR_DELETING_OPTIONAL_SERVICE = "Error deleting optional service \"{0}\" from offering \"{1}\" and plan \"{2}\": {3}";
     public static final String ERROR_MONITORING_CREATION_OF_SERVICES = "Error monitoring creation of services";
+    public static final String ERROR_MAPPING_SERVICE_NAME_TO_GUID = "Error mapping service name to guid: {0}";
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/message/Messages.java]: Please, follow the convention for writing customer-friendly messages :)

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/CreateOrUpdateServicesStep.java [hunk 5]
```diff
@@ -64,6 +65,9 @@
 
     @Inject
     private ServiceInstanceGetter serviceInstanceGetter;
+    
+    @Inject
+    private EventsGetter eventsGetter;
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/CreateOrUpdateServicesStep.java]: EventsGetter of what? Services, Applications?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteServicesStep.java [hunk 6]
```diff
@@ -59,6 +74,50 @@ protected StepPhase executeAsyncStep(ExecutionWrapper execution) throws Exceptio
...
+		String spaceId = StepsUtil.getSpaceId(execution.getContext());
+		CloudControllerClient client = execution.getControllerClient();
+		
+		Map<String, String> serviceNameGuidMap = services.parallelStream()
+			.collect(Collectors.toMap(Function.identity(), name -> {
+				
+				Map<String, Object> serviceInstanceWithMetadata = serviceInstanceGetter.getServiceInstanceWithMetadata(client, name, spaceId);
+				
+				@SuppressWarnings("unchecked")
+				String guid = (String) Stream.of(serviceInstanceWithMetadata)
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteServicesStep.java]: Pls, move from line 102-111 to ServiceInstanceGetter?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteServicesStep.java [hunk 7]
```diff
@@ -59,6 +74,50 @@ protected StepPhase executeAsyncStep(ExecutionWrapper execution) throws Exceptio
...
+			throw new SLException(Messages.ERROR_MAPPING_SERVICE_NAMES_TO_GUIDS);
+		}
+	}
+
+	private Map<String, String> getServiceGuids( List<String> services, ExecutionWrapper execution) {
+		String spaceId = StepsUtil.getSpaceId(execution.getContext());
+		CloudControllerClient client = execution.getControllerClient();
+		
+		Map<String, String> serviceNameGuidMap = services.parallelStream()
+			.collect(Collectors.toMap(Function.identity(), name -> {
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteServicesStep.java]: Too much nesting of the Streams...

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java [hunk 8]
```diff
@@ -107,11 +115,20 @@ public AsyncExecutionState execute(ExecutionWrapper execution) {
 
     private List<CloudServiceExtended> computeServicesToPoll(DelegateExecution context,
         Map<String, ServiceOperationType> triggeredServiceOperations) {
-        List<CloudServiceExtended> services = StepsUtil.getServicesToCreate(context);
+
+        List<String> servicesToDeleteNames = StepsUtil.getServicesToDelete(context);
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java]: What about refactoring of the PollServiceOperations... ?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java [hunk 9]
```diff
@@ -122,13 +139,30 @@ private ServiceOperation getLastServiceOperation(ExecutionWrapper execution, Clo
...
 
         return getLastOperation(execution, cloudServiceInstance);
     }
 
+    private boolean isServiceDeleted(ExecutionWrapper execution, UUID uuid) {
+        List<CloudEvent> serviceEvents = eventsGetter.getEvents(uuid);
+        if (serviceEvents != null) {
+            boolean isDeleted = serviceEvents.stream()
+                .filter(e -> e.getTimestamp() != null)
+                .sorted((a, b) -> b.getTimestamp()
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java]: Use order-by in /v2/events....

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/StepsUtil.java [hunk 10]
```diff
@@ -319,9 +319,10 @@ static void setCustomDomains(DelegateExecution context, List<String> customDomai
     @SuppressWarnings("unchecked")
     public static List<CloudServiceExtended> getServicesToCreate(DelegateExecution context) {
         List<String> services = (List<String>) context.getVariable(Constants.VAR_SERVICES_TO_CREATE);
-        return services.stream()
-            .map(service -> (CloudServiceExtended) JsonUtil.fromJson(service, CloudServiceExtended.class))
-            .collect(Collectors.toList());
+        return services == null ? new ArrayList<>()
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/StepsUtil.java]: Why not Collections.emptyList()?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/StepsUtil.java [hunk 11]
```diff
@@ -480,7 +480,8 @@ static void setAppsToUndeploy(DelegateExecution context, List<CloudApplication>
     }
 
     public static List<String> getServicesToDelete(DelegateExecution context) {
-        return getArrayVariableAsList(context, Constants.VAR_SERVICES_TO_DELETE);
+        List<String> arrayVariableAsList = getArrayVariableAsList(context, Constants.VAR_SERVICES_TO_DELETE);
+        return arrayVariableAsList != null ? arrayVariableAsList : new ArrayList<>();
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/StepsUtil.java]: Collections.emptyList();

File: com.sap.cloud.lm.sl.cf.process/src/test/java/com/sap/cloud/lm/sl/cf/process/steps/CreateOrUpdateServicesStepTest.java [hunk 12]
```diff
@@ -160,6 +160,7 @@ private void loadParameters() throws Exception {
 
     private void prepareContext() throws Exception {
         StepsUtil.setServicesToCreate(context, ListUtil.upcastUnmodifiable(stepInput.services));
+        StepsUtil.setServicesToDelete(context, new ArrayList<>());
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/test/java/com/sap/cloud/lm/sl/cf/process/steps/CreateOrUpdateServicesStepTest.java]: same

File: com.sap.cloud.lm.sl.cf.process/src/test/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsStepTest.java [hunk 13]
```diff
@@ -101,6 +103,8 @@ public void setUp() {
         context.setVariable(com.sap.cloud.lm.sl.cf.persistence.message.Constants.VARIABLE_NAME_SPACE_ID, TEST_SPACE_ID);
         prepareServiceInstanceGetter();
         StepsUtil.setServicesToCreate(context, input.services);
+        StepsUtil.setServicesToDelete(context, new ArrayList<>());
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/test/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsStepTest.java]: Same....Same

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java [hunk 14]
```diff
@@ -139,17 +139,27 @@ private ServiceOperation getLastServiceOperation(ExecutionWrapper execution, Clo
...
         if (cloudServiceInstance == null || cloudServiceInstance.isEmpty()) {
-            if (service.getMeta() != null) {
-                boolean isServiceDeleted = isServiceDeleted(execution, service.getMeta()
-                    .getGuid());
-                if (isServiceDeleted) {
-                    return new ServiceOperation(ServiceOperationType.DELETE, "", ServiceOperationState.SUCCEEDED);
-                }
+            ServiceOperation deleteOperation = getLastDeleteServiceOperation(execution, service);
+            if (deleteOperation == null) {
+                validateNullServiceInstance(execution, service);
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java]: validateNullServiceInstance o.O o.O

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java [hunk 15]
```diff
@@ -168,13 +178,12 @@ private boolean isServiceDeleted(ExecutionWrapper execution, UUID uuid) {
         return false;
     }
 
-    private void validateCloudServiceInstance(ExecutionWrapper execution, CloudServiceExtended service,
-        Map<String, Object> cloudServiceInstance) {
-        if (cloudServiceInstance == null && !service.isOptional()) {
+    private void validateNullServiceInstance(ExecutionWrapper execution, CloudServiceExtended service) {
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java]: Please explode this method in the .. ^^^^

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java [hunk 16]
```diff
@@ -165,19 +176,14 @@ private ServiceOperation getLastDeleteServiceOperation(ExecutionWrapper executio
...
         if (serviceEvents != null) {
-            return serviceEvents.stream()
-                .filter(e -> e.getTimestamp() != null)
-                .sorted((a, b) -> b.getTimestamp()
-                    .compareTo(a.getTimestamp()))
-                .findFirst()
-                .filter(e -> e.getType()
-                    .equals(SERVICE_EVENT_TYPE_DELETE))
-                .isPresent();
+            return serviceEvents.get(serviceEvents.size() - 1)
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java]: Not like:
1. serviceEvents.get(serviceEvents.size() - 1) could return null -> NPE => SERVICE_EVENT_TYPE_DELETE.equals(serviceEvents.get(serviceEvents.size() - 1));
2. Order by ... DESC returned from the CC ?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteServicesStep.java [hunk 17]
```diff
@@ -96,63 +97,53 @@ private void validateGuidMapping(Map<String, String> serviceGuids) {
...
-                    .filter(Objects::nonNull)
-                    .findFirst()
-                    .orElse("");
-
-                return guid;
-
-            }));
+            .collect(Collectors.toMap(Function.identity(), name -> this.serviceNameToGuid(name, client, spaceId)));
 
         serviceNameGuidMap.replaceAll((k, v) -> v.equals("") ? null : v);
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteServicesStep.java]: Could we remove this line? - I am not very happy with the ""

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteServicesStep.java [hunk 18]
```diff
@@ -96,63 +97,53 @@ private void validateGuidMapping(Map<String, String> serviceGuids) {
...
-                    .filter(Objects::nonNull)
-                    .map(map -> map.get("guid"))
-                    .filter(Objects::nonNull)
-                    .findFirst()
-                    .orElse("");
-
-                return guid;
-
-            }));
+            .collect(Collectors.toMap(Function.identity(), name -> this.serviceNameToGuid(name, client, spaceId)));
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteServicesStep.java]: Could we do:
1. Find all the serviceWithMetadata for each service
2. Filter the ones which have empty guid
3. map them to the corresponding service?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java [hunk 19]
```diff
@@ -116,20 +116,31 @@ public AsyncExecutionState execute(ExecutionWrapper execution) {
...
+    }
 
+    private List<CloudServiceExtended> getServicesToCreate(DelegateExecution context) {
+        List<CloudServiceExtended> allServicesToCreate = StepsUtil.getServicesToCreate(context);
         // There's no need to poll the creation or update of user-provided services, because it is done synchronously:
-        Stream<CloudServiceExtended> nonUserServicesToCreate = servicesToCreate.stream()
-            .filter(s -> !s.isUserProvided());
-        return Stream.concat(nonUserServicesToCreate, servicesToDelete.stream())
-            .filter(service -> triggeredServiceOperations.containsKey(service.getName()))
+        List<CloudServiceExtended> nonUserServicesToCreate = allServicesToCreate.stream()
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java]: return directly

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java [hunk 20]
```diff
@@ -116,20 +116,31 @@ public AsyncExecutionState execute(ExecutionWrapper execution) {
...
+            .collect(Collectors.toList());
+        return nonUserServicesToCreate;
+    }
+
+    private List<CloudServiceExtended> getServicesToDelete(DelegateExecution context) {
+        List<String> servicesToDeleteNames = StepsUtil.getServicesToDelete(context);
+        Map<String, String> servicesGuids = StepsUtil.getServicesGuids(context);
+
+        List<CloudServiceExtended> servicesToDelete = servicesToDeleteNames.stream()
+            .map(name -> new CloudServiceExtended(new Meta(UUID.fromString(servicesGuids.get(name)), null, null), name))
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java]: Why new Meta(UUID.fromString(servicesGuids.get(name))? Do you need this guid?
Could it be some random?

Comment (petko-boyadzhiev) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java]: We need it in the cloudServiceInstance

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java [hunk 21]
```diff
@@ -107,11 +114,38 @@ public AsyncExecutionState execute(ExecutionWrapper execution) {
 
     private List<CloudServiceExtended> computeServicesToPoll(DelegateExecution context,
         Map<String, ServiceOperationType> triggeredServiceOperations) {
-        List<CloudServiceExtended> services = StepsUtil.getServicesToCreate(context);
-        // There's no need to poll the creation or update of user-provided services, because it is done synchronously:
+
+        List<CloudServiceExtended> servicesToDelete = getServicesToDelete(context);
+        List<CloudServiceExtended> nonUserServicesToCreate = getServicesToCreate(context);
```
[the remark this answers is not in the document]
Comment (petko-boyadzhiev) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceOperationsExecution.java]: Changed to nonUserProvidedServicesToCreate

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteServicesStep.java [hunk 22]
```diff
@@ -17,63 +26,142 @@
...
         getStepLogger().info(Messages.DELETING_SERVICE, serviceName);
         client.deleteService(serviceName);
         getStepLogger().debug(Messages.SERVICE_DELETED, serviceName);
     }
 
+    private boolean validateServiceHasNoBindings(CloudServiceInstance serviceInstance, String serviceName) {
+        List<CloudServiceBinding> bindings = serviceInstance.getBindings();
+        if (!CollectionUtils.isEmpty(bindings)) {
+            logBindings(bindings);
+            getStepLogger().info(Messages.SERVICE_HAS_BINDINGS_AND_CANNOT_BE_DELETED, serviceName);
```
[the remark this answers is not in the document]
Comment (petko-boyadzhiev) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteServicesStep.java]: To show all services with issues resolving to guid

File: com.sap.cloud.lm.sl.cf.api/src/main/java/com/sap/cloud/lm/sl/cf/web/api/model/ProcessTypeJsonAdapter.java [hunk 23]
```diff
@@ -8,6 +8,8 @@
 
 public class ProcessTypeJsonAdapter extends TypeAdapter<ProcessType> {
 
+    private static final String NULL_STRING = (String) null;
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.api/src/main/java/com/sap/cloud/lm/sl/cf/web/api/model/ProcessTypeJsonAdapter.java]: Why not DEFAULT_NULL_VALUE or something similar?

File: com.sap.cloud.lm.sl.cf.api/src/main/java/com/sap/cloud/lm/sl/cf/web/api/model/ZonedDateTimeJsonAdapter.java [hunk 24]
```diff
@@ -19,7 +21,7 @@ public void write(JsonWriter out, ZonedDateTime zonedDateTime) throws IOExceptio
         if(zonedDateTime != null) {
             out.value(Operation.DATE_TIME_FORMATTER.format(zonedDateTime));            
         } else {
-            out.value((String) null);
+            out.value(NULL_STRING);
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.api/src/main/java/com/sap/cloud/lm/sl/cf/web/api/model/ZonedDateTimeJsonAdapter.java]: Same?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteServicesStep.java [hunk 25]
```diff
@@ -93,17 +86,14 @@ private void getServicesGuids(List<String> servicesToDeleteNames, ExecutionWrapp
...
+        List<String> servicesWithoutGuid = servicesToDelete.stream()
             .filter(name -> serviceGuids.get(name) == null)
-            .forEach(invalidEntry -> execution.getStepLogger()
-                .error(Messages.ERROR_MAPPING_SERVICE_NAME_TO_GUID, invalidEntry));
+            .collect(Collectors.toList());
 
-        Optional<String> firstServiceWithoutGuid = servicesToDelete.stream()
-            .filter(name -> serviceGuids.get(name) == null)
-            .findFirst();
+        String names = String.join(",", servicesWithoutGuid);
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteServicesStep.java]: Could you remove this?
And on line 96 to do: throw new SLException(Messages.ERROR_MAPPING_SERVICE_NAMES_TO_GUIDS, String.join(",", servicesWithoutGuid););

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/clients/ServiceGetter.java [hunk 26]
```diff
@@ -2,35 +2,39 @@
...
     private AbstractServiceGetter serviceInstanceGetter;
-    
-    @Autowired
-    @Qualifier("userProvidedServiceInstanceGetter")
     private AbstractServiceGetter userProvidedServiceInstanceGetter;
-    
+
+    @Inject
+    public ServiceGetter(@Qualifier("serviceInstanceGetter") AbstractServiceGetter serviceInstanceGetter,
+        @Qualifier("userProvidedServiceInstanceGetter") AbstractServiceGetter userProvidedServiceInstanceGetter) {
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/clients/ServiceGetter.java]: It is hard to read inline @Qualifier annotations. So, here you could inject the exact type??

File: com.sap.cloud.lm.sl.cf.api/src/main/java/com/sap/cloud/lm/sl/cf/web/api/model/ProcessTypeJsonAdapter.java (hunk 23 above)
Comment (petko-boyadzhiev) [reply com.sap.cloud.lm.sl.cf.api/src/main/java/com/sap/cloud/lm/sl/cf/web/api/model/ProcessTypeJsonAdapter.java]: It is a String reference with a null value. There can be Boolean with null too. We use it to suggest which overloaded method to be called, so for readability the type of the reference should be present in the variable name. We can use STRING_NULL_VALUE.

File: com.sap.cloud.lm.sl.cf.api/src/main/java/com/sap/cloud/lm/sl/cf/web/api/model/ZonedDateTimeJsonAdapter.java (hunk 24 above)
Comment (petko-boyadzhiev) [reply com.sap.cloud.lm.sl.cf.api/src/main/java/com/sap/cloud/lm/sl/cf/web/api/model/ZonedDateTimeJsonAdapter.java]: Same.

**Pull request merged**